### ISSUE TITLE
Attaching a collection to a container does not trigger the re-evaluation of its rules

### ISSUE THREAD

**Issue by dchristidis** [rucio/rucio#6234]
### Description

This appears to be a regression from #4611. It seems to have been partially fixed in #5676 but only for the implementation without temporary tables.

One thing that needs to be investigated is why the related test is successful.

On the operations side, it might be necessary to manually trigger the re-evaluation of all rules on containers.

### Steps to reproduce

1. Attach a dataset to a container that has a rule in `OK` state.
2. Watch the rule remain unaffected.

**Comment by bari12** [comment]
Needs to be backported to 1.29

**Comment by bari12** [comment]
Backport 1.29.13

**Comment by dchristidis** [comment]
On the operations side, in order to ensure that all rules are consistent, it is necessary to trigger their re-evaluation. All communities having used Rucio 1.27 and later are potentially affected.

 1. Update to a supported Rucio release where the regression is fixed (1.29.13 and 1.31.6 are the safest; some releases in the middle had a fix unless the temporary-table implementation was used.
 2. Due to the fact that containers can contain other containers, it is not seem possible to identify which ones have had their content changed in a specific time period. As such, it will be necessary to process almost all rules on containers. A direct database query is most efficient. The following has been tested on Oracle:
    ```sql
    SELECT RAWTOHEX(id)
    FROM rules
    WHERE
        did_type = 'C'
        AND state IN ('O', 'R')
    ````
    `STUCK` and `SUSPENDED` rules can be safely ignored. This query has be only run only once, the output stored in some file, then the step 3 run for subsets of the file (e.g. `split --lines 10000 ORIGINAL_FILE`).
 3. The most efficient way to re-evaluate them appears to be through the Conveyor Repairer. This mean executing the equivalent of `rucio update-rule --stuck RULE_ID` for each potentially-affected rule. The recommendation is to use the API layer from within Rucio and doing it in bunches, while monitoring the daemon’s backlog. This is quite community-specific (e.g. average size of containers, number of daemon instances). Using the Python client API:
    ```python
    from rucio.client import Client
    client = Client(account='root')    # or any other account with the required privileges
    with open('rules') as f:
        for line in f:
            client.update_replication_rule(rule_id=line.strip(), options={'state': 'STUCK'})
    ```
    or with the internal core layer (more efficient):
    ```python
    from rucio.core.rule import update_rule
    with open('rules') as f:
        for line in f:
            update_rule(rule_id=line.strip(), options={'state': 'STUCK'}
    ````
